# Segmenting and Clustering Neighbourhoods in Toronto

### This notebook contains 3 Parts of the Assignment

# Part 1

### Importing required libraries

In [3]:
import pandas as pd
import numpy as np
import requests
from bs4 import BeautifulSoup

Getting required table

In [39]:
# Get the HTML page of Wiki, and using read_html we convert the html data into list of Data frame objects
wiki_url = 'https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M'
response = requests.get(wiki_url)
soup = BeautifulSoup(response.text, 'lxml')
table_contents=[]
table=soup.find('table')
for row in table.findAll('td'):
    cell = {}
    if row.span.text=='Not assigned':
        pass
    else:
        cell['Postal Code'] = row.p.text[:3]
        cell['Borough'] = (row.span.text).split('(')[0]
        cell['Neighborhood'] = (((((row.span.text).split('(')[1]).strip(')')).replace(' /',',')).replace(')',' ')).strip(' ')
        table_contents.append(cell)

# print(table_contents)
df=pd.DataFrame(table_contents)
df['Borough']=df['Borough'].replace({'Downtown TorontoStn A PO Boxes25 The Esplanade':'Downtown Toronto Stn A',
                                             'East TorontoBusiness reply mail Processing Centre969 Eastern':'East Toronto Business',
                                             'EtobicokeNorthwest':'Etobicoke Northwest','East YorkEast Toronto':'East York/East Toronto',
                                             'MississaugaCanada Post Gateway Processing Centre':'Mississauga'})

Dropping Borough which are not assigned

In [40]:
#Remove cells which have borrow not assigned
filt = df['Neighborhood'] != 'Not assigned'
df = df[filt]
df.reset_index(inplace = True)
df.head()

index Postal Code           Borough                      Neighborhood
0      0         M3A        North York                         Parkwoods
1      1         M4A        North York                  Victoria Village
2      2         M5A  Downtown Toronto         Regent Park, Harbourfront
3      3         M6A        North York  Lawrence Manor, Lawrence Heights
4      4         M7A      Queen's Park     Ontario Provincial Government

In [41]:
df.groupby(['Postal Code']).head()

index Postal Code                Borough  \
0        0         M3A             North York   
1        1         M4A             North York   
2        2         M5A       Downtown Toronto   
3        3         M6A             North York   
4        4         M7A           Queen's Park   
..     ...         ...                    ...   
98      98         M8X              Etobicoke   
99      99         M4Y       Downtown Toronto   
100    100         M7Y  East Toronto Business   
101    101         M8Y              Etobicoke   
102    102         M8Z              Etobicoke   

                                          Neighborhood  
0                                            Parkwoods  
1                                     Victoria Village  
2                            Regent Park, Harbourfront  
3                     Lawrence Manor, Lawrence Heights  
4                        Ontario Provincial Government  
..                                                 ...  
98       The Kingsway, Montgomery Road, Old Mill North  
99                                Church and Wellesley  
100                                     Enclave of M4L  
101  Old Mill South, King's Mill Park, Sunnylea, Hu...  
102  Mimico NW, The Queensway West, South of Bloor,...  

[103 rows x 4 columns]

Number of records where Neighbourhood is "Not assigned"

In [43]:
df['Neighborhood'].str.count("Not assigned").sum()

0

In [44]:
df[df['Borough'] == 'Not assigned']

Empty DataFrame
Columns: [index, Postal Code, Borough, Neighborhood]
Index: []

In [45]:
df.shape

(103, 4)

# Part 2

Installing geocoder

In [46]:
pip install geocoder

Defaulting to user installation because normal site-packages is not writeable
You should consider upgrading via the '/usr/bin/python3 -m pip install --upgrade pip' command.
Note: you may need to restart the kernel to use updated packages.


In [47]:
import geocoder

Using csv since was running into errors with geocoder

In [48]:
url = 'https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-DS0701EN-SkillsNetwork/labs_v1/Geospatial_Coordinates.csv'
df_geo = pd.read_csv(url)
df_geo.head()

Postal Code   Latitude  Longitude
0         M1B  43.806686 -79.194353
1         M1C  43.784535 -79.160497
2         M1E  43.763573 -79.188711
3         M1G  43.770992 -79.216917
4         M1H  43.773136 -79.239476

In [49]:
df_geo.dtypes

Postal Code     object
Latitude       float64
Longitude      float64
dtype: object

In [50]:
df.shape

(103, 4)

In [51]:
df_geo.shape

(103, 3)

Combining both dataframes

In [52]:
df = df.join(df_geo.set_index('Postal Code'), on='Postal Code')
df

index Postal Code                Borough  \
0        0         M3A             North York   
1        1         M4A             North York   
2        2         M5A       Downtown Toronto   
3        3         M6A             North York   
4        4         M7A           Queen's Park   
..     ...         ...                    ...   
98      98         M8X              Etobicoke   
99      99         M4Y       Downtown Toronto   
100    100         M7Y  East Toronto Business   
101    101         M8Y              Etobicoke   
102    102         M8Z              Etobicoke   

                                          Neighborhood   Latitude  Longitude  
0                                            Parkwoods  43.753259 -79.329656  
1                                     Victoria Village  43.725882 -79.315572  
2                            Regent Park, Harbourfront  43.654260 -79.360636  
3                     Lawrence Manor, Lawrence Heights  43.718518 -79.464763  
4                        Ontario Provincial Government  43.662301 -79.389494  
..                                                 ...        ...        ...  
98       The Kingsway, Montgomery Road, Old Mill North  43.653654 -79.506944  
99                                Church and Wellesley  43.665860 -79.383160  
100                                     Enclave of M4L  43.662744 -79.321558  
101  Old Mill South, King's Mill Park, Sunnylea, Hu...  43.636258 -79.498509  
102  Mimico NW, The Queensway West, South of Bloor,...  43.628841 -79.520999  

[103 rows x 6 columns]

Resetting index

In [54]:
df = df.reset_index()

In [55]:
df.drop(['index'], axis = 'columns', inplace = True)
df = df.set_index('level_0')
df.head()

Postal Code           Borough                      Neighborhood  \
level_0                                                                   
0               M3A        North York                         Parkwoods   
1               M4A        North York                  Victoria Village   
2               M5A  Downtown Toronto         Regent Park, Harbourfront   
3               M6A        North York  Lawrence Manor, Lawrence Heights   
4               M7A      Queen's Park     Ontario Provincial Government   

          Latitude  Longitude  
level_0                        
0        43.753259 -79.329656  
1        43.725882 -79.315572  
2        43.654260 -79.360636  
3        43.718518 -79.464763  
4        43.662301 -79.389494

In [56]:
# df.rename(index = {'level_0' : 'index'})
df.index.name = 'index'
df.head()

Postal Code           Borough                      Neighborhood  \
index                                                                   
0             M3A        North York                         Parkwoods   
1             M4A        North York                  Victoria Village   
2             M5A  Downtown Toronto         Regent Park, Harbourfront   
3             M6A        North York  Lawrence Manor, Lawrence Heights   
4             M7A      Queen's Park     Ontario Provincial Government   

        Latitude  Longitude  
index                        
0      43.753259 -79.329656  
1      43.725882 -79.315572  
2      43.654260 -79.360636  
3      43.718518 -79.464763  
4      43.662301 -79.389494

In [57]:
df.shape

(103, 5)

# Part 3

Installing and folium for import

In [58]:
pip install geopy folium

Defaulting to user installation because normal site-packages is not writeable
You should consider upgrading via the '/usr/bin/python3 -m pip install --upgrade pip' command.
Note: you may need to restart the kernel to use updated packages.


In [59]:
# Using the foursquere API to segment and cluster the neighborhoods of Toronto
import geocoder
from geopy.geocoders import Nominatim 

address = 'Toronto, Ontario'

geolocator = Nominatim(user_agent="toronto_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of Toronto are {}, {}.'.format(latitude, longitude))

The geograpical coordinate of Toronto are 43.6534817, -79.3839347.


Creating Map of Toronto

In [60]:
import folium

# create map of Toronto using latitude and longitude values
map_Toronto = folium.Map(location=[latitude, longitude], zoom_start=10)

# add markers to map
for lat, lng, borough, neighborhood in zip(df['Latitude'], df['Longitude'], df['Borough'], df['Neighborhood']):
    label = '{}, {}'.format(neighborhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        ).add_to(map_Toronto)  
    
map_Toronto

Initializing Foursquare API credentials

In [61]:
CLIENT_ID = 'GS2R3ILSFSS0DYFJWFCTU12B53BBKBUUTGHWKFQIZDQZRTOB' 
CLIENT_SECRET = 'A4YVTHZGBR0YD1T52QGXJ1HITIWENIJD31UHDCWU1AEBSURB'
VERSION = '20180605' # Foursquare API version

print('Your credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)

Your credentails:
CLIENT_ID: GS2R3ILSFSS0DYFJWFCTU12B53BBKBUUTGHWKFQIZDQZRTOB
CLIENT_SECRET:A4YVTHZGBR0YD1T52QGXJ1HITIWENIJD31UHDCWU1AEBSURB


Next, we create a function to get all the venue categories in Toronto

In [86]:
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius
            )
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Category']
    
    return(nearby_venues)

In [87]:
venues_in_toronto = getNearbyVenues(df['Neighborhood'], df['Latitude'], df['Longitude'])

Parkwoods
Victoria Village
Regent Park, Harbourfront
Lawrence Manor, Lawrence Heights
Ontario Provincial Government
Islington Avenue
Malvern, Rouge
Don Mills North
Parkview Hill, Woodbine Gardens
Garden District, Ryerson
Glencairn
West Deane Park, Princess Gardens, Martin Grove, Islington, Cloverdale
Rouge Hill, Port Union, Highland Creek
Don Mills South
Woodbine Heights
St. James Town
Humewood-Cedarvale
Eringate, Bloordale Gardens, Old Burnhamthorpe, Markland Wood
Guildwood, Morningside, West Hill
The Beaches
Berczy Park
Caledonia-Fairbanks
Woburn
Leaside
Central Bay Street
Christie
Cedarbrae
Hillcrest Village
Bathurst Manor, Wilson Heights, Downsview North
Thorncliffe Park
Richmond, Adelaide, King
Dufferin, Dovercourt Village
Scarborough Village
Fairview, Henry Farm, Oriole
Northwood Park, York University
The Danforth  East
Harbourfront East, Union Station, Toronto Islands
Little Portugal, Trinity
Kennedy Park, Ionview, East Birchmount Park
Bayview Village
Downsview East
The Danforth

In [88]:
venues_in_toronto.shape

(1339, 5)

In [89]:
venues_in_toronto.head()

Neighborhood  Neighborhood Latitude  Neighborhood Longitude  \
0         Parkwoods              43.753259              -79.329656   
1         Parkwoods              43.753259              -79.329656   
2         Parkwoods              43.753259              -79.329656   
3  Victoria Village              43.725882              -79.315572   
4  Victoria Village              43.725882              -79.315572   

                    Venue        Venue Category  
0         Brookbanks Park                  Park  
1                     KFC  Fast Food Restaurant  
2           Variety Store     Food & Drink Shop  
3  Victoria Village Arena          Hockey Arena  
4             Tim Hortons           Coffee Shop

Checking the Venues based on Neighbourhood

In [90]:
venues_in_toronto.groupby('Neighborhood').head()

Neighborhood  \
0                                             Parkwoods   
1                                             Parkwoods   
2                                             Parkwoods   
3                                      Victoria Village   
4                                      Victoria Village   
...                                                 ...   
1326  Mimico NW, The Queensway West, South of Bloor,...   
1327  Mimico NW, The Queensway West, South of Bloor,...   
1328  Mimico NW, The Queensway West, South of Bloor,...   
1329  Mimico NW, The Queensway West, South of Bloor,...   
1330  Mimico NW, The Queensway West, South of Bloor,...   

      Neighborhood Latitude  Neighborhood Longitude                   Venue  \
0                 43.753259              -79.329656         Brookbanks Park   
1                 43.753259              -79.329656                     KFC   
2                 43.753259              -79.329656           Variety Store   
3                 43.725882              -79.315572  Victoria Village Arena   
4                 43.725882              -79.315572             Tim Hortons   
...                     ...                     ...                     ...   
1326              43.628841              -79.520999        South St. Burger   
1327              43.628841              -79.520999              Wingporium   
1328              43.628841              -79.520999               Dollarama   
1329              43.628841              -79.520999          Healthy Planet   
1330              43.628841              -79.520999    Artisano Bakery Café   

            Venue Category  
0                     Park  
1     Fast Food Restaurant  
2        Food & Drink Shop  
3             Hockey Arena  
4              Coffee Shop  
...                    ...  
1326          Burger Joint  
1327           Wings Joint  
1328        Discount Store  
1329       Supplement Shop  
1330                Bakery  

[430 rows x 5 columns]

Checking for the maximum venue categories

In [91]:
venues_in_toronto.groupby('Venue Category').max()

Neighborhood  \
Venue Category                                                          
Accessories Store                    Lawrence Manor, Lawrence Heights   
Adult Boutique                                   Church and Wellesley   
Airport                                                Downsview East   
Airport Food Court  CN Tower, King and Spadina, Railway Lands, Har...   
Airport Gate        CN Tower, King and Spadina, Railway Lands, Har...   
...                                                               ...   
Warehouse Store                                      Thorncliffe Park   
Wine Bar                                     Little Portugal, Trinity   
Wings Joint         Mimico NW, The Queensway West, South of Bloor,...   
Women's Store                        Lawrence Manor, Lawrence Heights   
Yoga Studio                            University of Toronto, Harbord   

                    Neighborhood Latitude  Neighborhood Longitude  \
Venue Category                                                      
Accessories Store               43.718518              -79.464763   
Adult Boutique                  43.665860              -79.383160   
Airport                         43.737473              -79.394420   
Airport Food Court              43.628947              -79.394420   
Airport Gate                    43.628947              -79.394420   
...                                   ...                     ...   
Warehouse Store                 43.705369              -79.349372   
Wine Bar                        43.653206              -79.400049   
Wings Joint                     43.628841              -79.520999   
Women's Store                   43.718518              -79.453512   
Yoga Studio                     43.715383              -79.321558   

                                              Venue  
Venue Category                                       
Accessories Store               Ardene Shoes Outlet  
Adult Boutique                            Seduction  
Airport             Toronto Downsview Airport (YZD)  
Airport Food Court                Billy Bishop Café  
Airport Gate                                 Gate 8  
...                                             ...  
Warehouse Store                              Costco  
Wine Bar                            Paris Paris Bar  
Wings Joint                              Wingporium  
Women's Store                         Maximum Woman  
Yoga Studio                               YogaSpace  

[230 rows x 4 columns]

There are around 230 different types of Venue Categories

### One Hot encoding the venue Categories

In [92]:
toronto_venue_cat = pd.get_dummies(venues_in_toronto[['Venue Category']], prefix="", prefix_sep="")
toronto_venue_cat

Accessories Store  Adult Boutique  Airport  Airport Food Court  \
0                     0               0        0                   0   
1                     0               0        0                   0   
2                     0               0        0                   0   
3                     0               0        0                   0   
4                     0               0        0                   0   
...                 ...             ...      ...                 ...   
1334                  0               0        0                   0   
1335                  0               0        0                   0   
1336                  0               0        0                   0   
1337                  0               0        0                   0   
1338                  0               0        0                   0   

      Airport Gate  Airport Lounge  Airport Service  Airport Terminal  \
0                0               0                0                 0   
1                0               0                0                 0   
2                0               0                0                 0   
3                0               0                0                 0   
4                0               0                0                 0   
...            ...             ...              ...               ...   
1334             0               0                0                 0   
1335             0               0                0                 0   
1336             0               0                0                 0   
1337             0               0                0                 0   
1338             0               0                0                 0   

      American Restaurant  Antique Shop  ...  Truck Stop  \
0                       0             0  ...           0   
1                       0             0  ...           0   
2                       0             0  ...           0   
3                       0             0  ...           0   
4                       0             0  ...           0   
...                   ...           ...  ...         ...   
1334                    0             0  ...           0   
1335                    0             0  ...           0   
1336                    0             0  ...           0   
1337                    0             0  ...           0   
1338                    0             0  ...           0   

      Vegetarian / Vegan Restaurant  Video Game Store  Video Store  \
0                                 0                 0            0   
1                                 0                 0            0   
2                                 0                 0            0   
3                                 0                 0            0   
4                                 0                 0            0   
...                             ...               ...          ...   
1334                              0                 0            0   
1335                              0                 0            0   
1336                              0                 0            0   
1337                              0                 0            0   
1338                              0                 0            0   

      Vietnamese Restaurant  Warehouse Store  Wine Bar  Wings Joint  \
0                         0                0         0            0   
1                         0                0         0            0   
2                         0                0         0            0   
3                         0                0         0            0   
4                         0                0         0            0   
...                     ...              ...       ...          ...   
1334                      0                0         0            0   
1335                      0                0         0            0   
1336                      0                0         0         

Adding the neighbourhood to the encoded dataframe

In [93]:
toronto_venue_cat['Neighborhood'] = venues_in_toronto['Neighborhood'] 

# moving neighborhood column to the first column
fixed_columns = [toronto_venue_cat.columns[-1]] + list(toronto_venue_cat.columns[:-1])
toronto_venue_cat = toronto_venue_cat[fixed_columns]

toronto_venue_cat.head()

Yoga Studio  Accessories Store  Adult Boutique  Airport  \
0            0                  0               0        0   
1            0                  0               0        0   
2            0                  0               0        0   
3            0                  0               0        0   
4            0                  0               0        0   

   Airport Food Court  Airport Gate  Airport Lounge  Airport Service  \
0                   0             0               0                0   
1                   0             0               0                0   
2                   0             0               0                0   
3                   0             0               0                0   
4                   0             0               0                0   

   Airport Terminal  American Restaurant  ...  Train Station  Truck Stop  \
0                 0                    0  ...              0           0   
1                 0                    0  ...              0           0   
2                 0                    0  ...              0           0   
3                 0                    0  ...              0           0   
4                 0                    0  ...              0           0   

   Vegetarian / Vegan Restaurant  Video Game Store  Video Store  \
0                              0                 0            0   
1                              0                 0            0   
2                              0                 0            0   
3                              0                 0            0   
4                              0                 0            0   

   Vietnamese Restaurant  Warehouse Store  Wine Bar  Wings Joint  \
0                      0                0         0            0   
1                      0                0         0            0   
2                      0                0         0            0   
3                      0                0         0            0   
4                      0                0         0            0   

   Women's Store  
0              0  
1              0  
2              0  
3              0  
4              0  

[5 rows x 230 columns]

We will group the Neighbourhoods, calculate the mean venue categories in each Neighbourhood

In [94]:
toronto_grouped = toronto_venue_cat.groupby('Neighborhood').mean().reset_index()
toronto_grouped.head()

Neighborhood  Yoga Studio  \
0                                        Agincourt          0.0   
1                           Alderwood, Long Branch          0.0   
2  Bathurst Manor, Wilson Heights, Downsview North          0.0   
3                                  Bayview Village          0.0   
4                Bedford Park, Lawrence Manor East          0.0   

   Accessories Store  Adult Boutique  Airport  Airport Food Court  \
0                0.0             0.0      0.0                 0.0   
1                0.0             0.0      0.0                 0.0   
2                0.0             0.0      0.0                 0.0   
3                0.0             0.0      0.0                 0.0   
4                0.0             0.0      0.0                 0.0   

   Airport Gate  Airport Lounge  Airport Service  Airport Terminal  ...  \
0           0.0             0.0              0.0               0.0  ...   
1           0.0             0.0              0.0               0.0  ...   
2           0.0             0.0              0.0               0.0  ...   
3           0.0             0.0              0.0               0.0  ...   
4           0.0             0.0              0.0               0.0  ...   

   Train Station  Truck Stop  Vegetarian / Vegan Restaurant  Video Game Store  \
0            0.0         0.0                            0.0               0.0   
1            0.0         0.0                            0.0               0.0   
2            0.0         0.0                            0.0               0.0   
3            0.0         0.0                            0.0               0.0   
4            0.0         0.0                            0.0               0.0   

   Video Store  Vietnamese Restaurant  Warehouse Store  Wine Bar  Wings Joint  \
0          0.0                    0.0              0.0       0.0          0.0   
1          0.0                    0.0              0.0       0.0          0.0   
2          0.0                    0.0              0.0       0.0          0.0   
3          0.0                    0.0              0.0       0.0          0.0   
4          0.0                    0.0              0.0       0.0          0.0   

   Women's Store  
0            0.0  
1            0.0  
2            0.0  
3            0.0  
4            0.0  

[5 rows x 230 columns]

Function to get the most common venue

In [95]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

In [96]:
import numpy as np

num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighborhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighborhood'] = toronto_grouped['Neighborhood']

for ind in np.arange(toronto_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(toronto_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()

Neighborhood      1st Most Common Venue  \
0                                        Agincourt  Latin American Restaurant   
1                           Alderwood, Long Branch                Pizza Place   
2  Bathurst Manor, Wilson Heights, Downsview North                       Bank   
3                                  Bayview Village                       Bank   
4                Bedford Park, Lawrence Manor East         Italian Restaurant   

  2nd Most Common Venue      3rd Most Common Venue 4th Most Common Venue  \
0                Lounge             Clothing Store        Breakfast Spot   
1           Coffee Shop                       Pool        Sandwich Place   
2           Coffee Shop  Middle Eastern Restaurant           Supermarket   
3    Chinese Restaurant        Japanese Restaurant                  Café   
4           Coffee Shop             Sandwich Place            Restaurant   

  5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0          Dance Studio             Drugstore            Donut Shop   
1              Pharmacy                   Gym                   Pub   
2            Restaurant      Sushi Restaurant          Intersection   
3         Women's Store          Dance Studio             Drugstore   
4                   Pub  Fast Food Restaurant               Butcher   

  8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0               Dog Run   Distribution Center         Discount Store  
1          Cupcake Shop           Curling Ice           Dance Studio  
2                 Diner         Deli / Bodega      Health Food Store  
3            Donut Shop               Dog Run    Distribution Center  
4      Sushi Restaurant                  Café      Indian Restaurant

Making the model to cluster our Neighbourhoods

In [97]:
# import k-means from clustering stage
from sklearn.cluster import KMeans

In [98]:
# set number of clusters
k_num_clusters = 5

toronto_grouped_clustering = toronto_grouped.drop('Neighborhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=k_num_clusters, random_state=0).fit(toronto_grouped_clustering)
kmeans

KMeans(n_clusters=5, random_state=0)

In [99]:
kmeans.labels_[0:100]

array([0, 0, 0, 0, 0, 0, 0, 0, 0, 2, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 3, 2, 0, 0, 0, 0, 0, 0, 0, 0, 0, 2, 0, 2, 0, 0, 0, 0, 0, 4,
       3, 0, 0, 0, 0, 2, 0, 2, 0, 0, 0, 2, 0, 2, 0, 2, 0, 0, 3, 0, 0, 0,
       2, 0, 0, 2, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 2, 0, 2, 0, 0,
       0, 0, 0, 0, 1, 0, 0, 0, 1, 0, 0, 2], dtype=int32)

In [100]:
neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

In [101]:
toronto_merged = df
toronto_merged = toronto_merged.join(neighborhoods_venues_sorted.set_index('Neighborhood'), on='Neighborhood')
toronto_merged.head()

Postal Code           Borough                      Neighborhood  \
index                                                                   
0             M3A        North York                         Parkwoods   
1             M4A        North York                  Victoria Village   
2             M5A  Downtown Toronto         Regent Park, Harbourfront   
3             M6A        North York  Lawrence Manor, Lawrence Heights   
4             M7A      Queen's Park     Ontario Provincial Government   

        Latitude  Longitude  Cluster Labels 1st Most Common Venue  \
index                                                               
0      43.753259 -79.329656             2.0                  Park   
1      43.725882 -79.315572             0.0           Coffee Shop   
2      43.654260 -79.360636             0.0           Coffee Shop   
3      43.718518 -79.464763             0.0        Clothing Store   
4      43.662301 -79.389494             0.0           Coffee Shop   

       2nd Most Common Venue 3rd Most Common Venue   4th Most Common Venue  \
index                                                                        
0       Fast Food Restaurant     Food & Drink Shop             Curling Ice   
1      Portuguese Restaurant          Hockey Arena             Pizza Place   
2                     Bakery                  Park          Breakfast Spot   
3              Women's Store   Arts & Crafts Store  Furniture / Home Store   
4           Sushi Restaurant           Yoga Studio          Sandwich Place   

      5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
index                                                                     
0                Donut Shop               Dog Run   Distribution Center   
1              Intersection     French Restaurant          Dance Studio   
2         French Restaurant         Historic Site        Farmers Market   
3                Shoe Store           Event Space   Sporting Goods Shop   
4                       Bar              Beer Bar                   Spa   

      8th Most Common Venue  9th Most Common Venue 10th Most Common Venue  
index                                                                      
0            Discount Store                  Diner     Dim Sum Restaurant  
1             Deli / Bodega       Department Store           Dessert Shop  
2       Distribution Center           Dessert Shop     Mexican Restaurant  
3               Coffee Shop  Vietnamese Restaurant               Boutique  
4        Italian Restaurant    Japanese Restaurant    Distribution Center

Drop all the NaN values to prevent data skew

In [102]:
toronto_merged_nonan = toronto_merged.dropna(subset=['Cluster Labels'])

In [103]:
import matplotlib.cm as cm
import matplotlib.colors as colors

Plotting the clusters on the map

In [105]:
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(k_num_clusters)
ys = [i + x + (i*x)**2 for i in range(k_num_clusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(toronto_merged_nonan['Latitude'], toronto_merged_nonan['Longitude'], toronto_merged_nonan['Neighborhood'], toronto_merged_nonan['Cluster Labels']):
    label = folium.Popup('Cluster ' + str(int(cluster) +1) + '\n' + str(poi) , parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[int(cluster-1)],
        fill=True,
        fill_color=rainbow[int(cluster-1)]
        ).add_to(map_clusters)
        
map_clusters

### Verifying each cluster

#### Cluster 1

In [106]:
toronto_merged_nonan.loc[toronto_merged_nonan['Cluster Labels'] == 0, toronto_merged_nonan.columns[[1] + list(range(5, toronto_merged_nonan.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
index                                                                
1                 North York             0.0           Coffee Shop   
2           Downtown Toronto             0.0           Coffee Shop   
3                 North York             0.0        Clothing Store   
4               Queen's Park             0.0           Coffee Shop   
6                Scarborough             0.0  Fast Food Restaurant   
...                      ...             ...                   ...   
96          Downtown Toronto             0.0                  Café   
97          Downtown Toronto             0.0            Restaurant   
99          Downtown Toronto             0.0      Sushi Restaurant   
100    East Toronto Business             0.0    Light Rail Station   
102                Etobicoke             0.0         Tanning Salon   

       2nd Most Common Venue 3rd Most Common Venue   4th Most Common Venue  \
index                                                                        
1      Portuguese Restaurant          Hockey Arena             Pizza Place   
2                     Bakery                  Park          Breakfast Spot   
3              Women's Store   Arts & Crafts Store  Furniture / Home Store   
4           Sushi Restaurant           Yoga Studio          Sandwich Place   
6              Women's Store           Curling Ice               Drugstore   
...                      ...                   ...                     ...   
96                    Bakery    Italian Restaurant              Restaurant   
97                      Café           Coffee Shop      Seafood Restaurant   
99               Pizza Place          Dance Studio             Escape Room   
100              Pizza Place            Skate Park                 Brewery   
102           Sandwich Place       Supplement Shop                  Bakery   

      5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
index                                                                     
1              Intersection     French Restaurant          Dance Studio   
2         French Restaurant         Historic Site        Farmers Market   
3                Shoe Store           Event Space   Sporting Goods Shop   
4                       Bar              Beer Bar                   Spa   
6                Donut Shop               Dog Run   Distribution Center   
...                     ...                   ...                   ...   
96               Playground       Thai Restaurant                  Bank   
97                    Hotel  Gym / Fitness Center   Japanese Restaurant   
99                      Pub      Ramen Restaurant        Ice Cream Shop   
100           Burrito Place        Farmers Market  Fast Food Restaurant   
102            Burger Joint  Fast Food Restaurant        Hardware Store   

      8th Most Common Venue  9th Most Common Venue 10th Most Common Venue  
index                                                                      
1             Deli / Bodega       Department Store           Dessert Shop  
2       Distribution Center           Dessert Shop     Mexican Restaurant  
3               Coffee Shop  Vietnamese Restaurant               Boutique  
4        Italian Restaurant    Japanese Restaurant    Distribution Center  
6            Discount Store                  Diner     Dim Sum Restaurant  
...                     ...                    ...                    ...  
96            Deli / Bodega                 Market             Beer Store  
97                Speakeasy                    Pub                 Bakery  
99                 Beer Bar      Indian Restaurant             Restaurant  
100              Restaurant          Auto Workshop            Yoga Studio  
102       Convenience Store                    Gym          Grocery Store  

[80 rows x 12 columns]

#### Cluster 2

In [107]:
toronto_merged_nonan.loc[toronto_merged_nonan['Cluster Labels'] == 1, toronto_merged_nonan.columns[[1] + list(range(5, toronto_merged_nonan.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
index                                                                           
52     North York             1.0                  Park     Electronics Store   
64           York             1.0                  Park     Electronics Store   

      3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
index                                                                     
52                Drugstore            Donut Shop               Dog Run   
64                Drugstore            Donut Shop               Dog Run   

      6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
index                                                                     
52      Distribution Center        Discount Store                 Diner   
64      Distribution Center        Discount Store                 Diner   

      9th Most Common Venue 10th Most Common Venue  
index                                               
52       Dim Sum Restaurant           Dessert Shop  
64       Dim Sum Restaurant           Dessert Shop

#### Cluster 3

In [108]:
toronto_merged_nonan.loc[toronto_merged_nonan['Cluster Labels'] == 2, toronto_merged_nonan.columns[[1] + list(range(5, toronto_merged_nonan.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
index                                                                 
0                  North York             2.0                  Park   
10                 North York             2.0                  Park   
21                       York             2.0                  Park   
35     East York/East Toronto             2.0          Intersection   
40                 North York             2.0                  Park   
49                 North York             2.0                  Park   
61            Central Toronto             2.0      Business Service   
66                 North York             2.0                  Park   
68            Central Toronto             2.0                  Park   
77                  Etobicoke             2.0                  Park   
83            Central Toronto             2.0                  Park   
85                Scarborough             2.0                  Park   
91           Downtown Toronto             2.0                  Park   
98                  Etobicoke             2.0                  Park   

      2nd Most Common Venue       3rd Most Common Venue 4th Most Common Venue  \
index                                                                           
0      Fast Food Restaurant           Food & Drink Shop           Curling Ice   
10      Japanese Restaurant           Convenience Store    Italian Restaurant   
21                     Pool               Women's Store               Gay Bar   
35                     Park           Convenience Store           Curling Ice   
40                  Airport                Cupcake Shop             Drugstore   
49                   Bakery  Construction & Landscaping        Massage Studio   
61                 Bus Line                 Swim School                  Park   
66        Convenience Store                 Curling Ice             Drugstore   
68            Jewelry Store                       Trail      Sushi Restaurant   
77        Mobile Phone Shop              Sandwich Place              Bus Line   
83                      Gym                       Trail          Cupcake Shop   
85               Playground                Intersection           Curling Ice   
91               Playground                       Trail      Cuban Restaurant   
98                     Pool                       River      Cuban Restaurant   

      5th Most Common Venue    6th Most Common Venue 7th Most Common Venue  \
index                                                                        
0                Donut Shop                  Dog Run   Distribution Center   
10                   Bakery            Women's Store         Deli / Bodega   
21                 Creperie      Distribution Center        Discount Store   
35                Drugstore               Donut Shop               Dog Run   
40               Donut Shop                  Dog Run   Distribution Center   
49              Curling Ice               Donut Shop               Dog Run   
61          College Stadium  Comfort Food Restaurant            Donut Shop   
66               Donut Shop                  Dog Run   Distribution Center   
68            Women's Store             Cupcake Shop               Dog Run   
77              Curling Ice               Donut Shop               Dog Run   
83                Drugstore               Donut Shop               Dog Run   
85                Drugstore               Donut Shop               Dog Run   
91               Donut Shop                  Dog Run   Distribution Center   
98               Donut Shop                  Dog Run   Distribution Center   

      8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
index                                                                     
0            Discount Store                 Diner     Dim Sum Restaurant  
10                Drugstore            Donut Shop                Dog Run  
21                    Diner    Dim Sum 

#### Cluster 4

In [109]:
toronto_merged_nonan.loc[toronto_merged_nonan['Cluster Labels'] == 3, toronto_merged_nonan.columns[[1] + list(range(5, toronto_merged_nonan.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
index                                                                           
53     North York             3.0     Korean Restaurant        Baseball Field   
57     North York             3.0        Baseball Field         Women's Store   
101     Etobicoke             3.0        Baseball Field                  Pool   

      3rd Most Common Venue        4th Most Common Venue  \
index                                                      
53               Food Truck                Women's Store   
57             Dance Studio  Eastern European Restaurant   
101           Women's Store                  Curling Ice   

      5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
index                                                                     
53        Electronics Store             Drugstore            Donut Shop   
57                Drugstore            Donut Shop               Dog Run   
101               Drugstore            Donut Shop               Dog Run   

      8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
index                                                                     
53                  Dog Run   Distribution Center         Discount Store  
57      Distribution Center        Discount Store                  Diner  
101     Distribution Center        Discount Store                  Diner

#### Cluster 5

In [110]:
toronto_merged_nonan.loc[toronto_merged_nonan['Cluster Labels'] == 4, toronto_merged_nonan.columns[[1] + list(range(5, toronto_merged_nonan.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
index                                                                           
50     North York             4.0                   Gym         Women's Store   

      3rd Most Common Venue        4th Most Common Venue  \
index                                                      
50              Curling Ice  Eastern European Restaurant   

      5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
index                                                                     
50                Drugstore            Donut Shop               Dog Run   

      8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
index                                                                     
50      Distribution Center        Discount Store                  Diner

Successfully cluster Toronto neighbourhood based on venue categories